### PR TITLE
Scratchpad refactoring

### LINKED ISSUE #1261: Scratchpad allocation / deallocation are spread in 2 files
Allocation antares/study/runtime/runtime.cpp:105

Deallocation antares/study/area/area.cpp:119

Use smart pointers / std::vectors instead.

### PR DESCRIPTION
**Pull request by payetvin**
close #1261

### CODE REVIEW

File: src/libs/antares/study/area/area.h [hunk 1]
```diff
@@ -321,7 +321,7 @@ class Area final : private Yuni::NonCopyable<Area>
     /*!
     ** \brief Scratchpad used temporary calculations (solver only)
     */
-    mutable AreaScratchpad** scratchpad;
+    mutable std::vector<std::shared_ptr<AreaScratchpad>> scratchpad;
```
Comment (flomnes) [inline src/libs/antares/study/area/area.h]:
```suggestion
    mutable std::vector<AreaScratchpad> scratchpad;
```

File: src/libs/antares/study/runtime/runtime.cpp [hunk 2]
```diff
@@ -100,9 +100,9 @@ static void StudyRuntimeInfosInitializeAllAreas(Study& study, StudyRuntimeInfos&
             area.thermal.mustrunList.calculationOfSpinning();
         }
 
-        area.scratchpad = new AreaScratchpad*[area.nbYearsInParallel];
+        area.scratchpad = std::vector<std::shared_ptr<AreaScratchpad>>(area.nbYearsInParallel);
```
Comment (flomnes) [inline src/libs/antares/study/runtime/runtime.cpp]:
```suggestion
        area.scratchpad.reserve(area.nbYearsInParallel);
```

File: src/libs/antares/study/runtime/runtime.cpp [hunk 3]
```diff
@@ -100,9 +100,9 @@ static void StudyRuntimeInfosInitializeAllAreas(Study& study, StudyRuntimeInfos&
...
         for (uint numSpace = 0; numSpace < area.nbYearsInParallel; numSpace++)
-            area.scratchpad[numSpace] = new AreaScratchpad(r, area);
+            area.scratchpad[numSpace] = std::make_shared<AreaScratchpad>(r, area);
```
Comment (flomnes) [inline src/libs/antares/study/runtime/runtime.cpp]:
```suggestion
            area.scratchpad.emplace_back(r, area);
```

File: src/solver/variable/state.hxx [hunk 4]
```diff
@@ -71,7 +71,7 @@ inline void State::yearEndResetThermal()
 inline void State::initFromAreaIndex(const unsigned int areaIndex, uint numSpace)
 {
     area = study.areas[areaIndex];
-    scratchpad = area->scratchpad[numSpace];
+    scratchpad = area->scratchpad[numSpace].get();
```
Comment (flomnes) [inline src/solver/variable/state.hxx]:
```suggestion
    scratchpad = &area->scratchpad[numSpace];
```

**Review by flomnes — CHANGES_REQUESTED** [review]
One of the `*` in `mutable AreaScratchpad** scratchpad` was useless

File: src/libs/antares/study/area/area.cpp [hunk 5]
```diff
@@ -239,8 +224,8 @@ Yuni::uint64 Area::memoryUsage() const
         ret += ui->memoryUsage();
 
     // scratchpad
-    if (scratchpad)
-        ret += sizeof(AreaScratchpad) * nbYearsInParallel;
+    if (!scratchpad.empty())
+        ret += sizeof(scratchpad[0]) * scratchpad.size();
```
Comment (flomnes) [inline src/libs/antares/study/area/area.cpp]: I think the former one was clearer

**Review by flomnes — APPROVED** [review]

File: src/libs/antares/study/area/scratchpad.cpp [hunk 6]
```diff
@@ -157,5 +157,7 @@ AreaScratchpad::AreaScratchpad(const StudyRuntimeInfos& rinfos, Area& area) : ts
     pumpHasMod = (valuePumping > 0.);
 }
 
+AreaScratchpad::~AreaScratchpad() = default;
```
Comment (guilpier-code) [inline src/libs/antares/study/area/scratchpad.cpp:160]: We should remove the destructor here, and make it a default one in the corresponding **.h** source file

**Review by guilpier-code — CHANGES_REQUESTED** [review]

File: src/libs/antares/study/area/scratchpad.cpp (hunk 6 above)
Comment (flomnes) [reply src/libs/antares/study/area/scratchpad.cpp:160]: Placing all declarations in the header, and all definitions in the .cpp is good practice. I'd rather keep the way @payetvin did.

**Review by guilpier-code — APPROVED** [review]

**Pull request merged**